File: apps/backend-java/src/main/java/com/housingplatform/hosts/HostRepository.java
package com.housingplatform.hosts;

import com.housingplatform.persistence.entity.Host;
import com.housingplatform.persistence.enums.BookingStatus;
import com.housingplatform.persistence.enums.BookingType;
import jakarta.persistence.EntityManager;
import jakarta.persistence.NoResultException;
import jakarta.persistence.PersistenceContext;
import java.time.LocalDate;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.UUID;
import org.springframework.stereotype.Repository;

@Repository
public class HostRepository {

  @PersistenceContext private EntityManager entityManager;

  public record HostPropertyListRow(
      UUID id,
      String title,
      String slug,
      String propertyType,
      String district,
      String status,
      String bookingMode,
      Integer monthlyPriceMin,
      int roomCount,
      OffsetDateTime updatedAt) {}

  public record HostBookingRow(
      UUID id,
      BookingStatus status,
      BookingType bookingType,
      LocalDate checkIn,
      LocalDate checkOut,
      int guestCount,
      String customerNotes,
      String propertyTitle,
      String roomName,
      int totalKrw,
      OffsetDateTime createdAt) {}

  public Optional<Host> findByProfileId(UUID profileId) {
    return entityManager
        .createQuery(
            """
            select h from Host h
            where h.profileId = :profileId and h.deletedAt is null
            """,
            Host.class)
        .setParameter("profileId", profileId)
        .getResultStream()
        .findFirst();
  }

  public Host register(UUID profileId, String displayName) {
    Optional<Host> existing = findByProfileId(profileId);
    if (existing.isPresent()) {
      return existing.get();
    }

    entityManager
        .createNativeQuery("alter table public.profiles disable trigger profiles_protect_role")
        .executeUpdate();
    try {
      entityManager
          .createNativeQuery(
              """
              update public.profiles set role = 'host' where id = :profileId
              """)
          .setParameter("profileId", profileId)
          .executeUpdate();

      UUID hostId =
          (UUID)
              entityManager
                  .createNativeQuery(
                      """
                      insert into public.hosts (profile_id, display_name, status)
                      values (:profileId, :displayName, 'pending')
                      returning id
                      """)
                  .setParameter("profileId", profileId)
                  .setParameter("displayName", displayName.strip())
                  .getSingleResult();
      entityManager.flush();
      return findById(hostId).orElseThrow();
    } finally {
      entityManager
          .createNativeQuery("alter table public.profiles enable trigger profiles_protect_role")
          .executeUpdate();
    }
  }

  public Optional<Host> findById(UUID hostId) {
    return entityManager
        .createQuery("select h from Host h where h.id = :hostId", Host.class)
        .setParameter("hostId", hostId)
        .getResultStream()
        .findFirst();
  }

  @SuppressWarnings("unchecked")
  public List<HostPropertyListRow> listProperties(UUID hostId) {
    List<Object[]> rows =
        entityManager
            .createNativeQuery(
                """
                select
                  p.id,
                  p.title,
                  p.slug,
                  p.property_type::text,
                  p.district,
                  p.status::text,
                  p.booking_mode::text,
                  p.monthly_price_min,
                  coalesce(
                    (
                      select count(*)::integer
                      from public.rooms r
                      where r.property_id = p.id and r.deleted_at is null
                    ),
                    0
                  ),
                  p.updated_at
                from public.properties p
                where p.host_id = :hostId and p.deleted_at is null
                order by p.updated_at desc
                """)
            .setParameter("hostId", hostId)
            .getResultList();

    List<HostPropertyListRow> results = new ArrayList<>();
    for (Object[] row : rows) {
      results.add(
          new HostPropertyListRow(
              (UUID) row[0],
              (String) row[1],
              (String) row[2],
              (String) row[3],
              (String) row[4],
              (String) row[5],
              (String) row[6],
              row[7] == null ? null : ((Number) row[7]).intValue(),
              ((Number) row[8]).intValue(),
              (OffsetDateTime) row[9]));
    }
    return results;
  }

  public boolean propertyBelongsToHost(UUID hostId, UUID propertyId) {
    Number count =
        (Number)
            entityManager
                .createNativeQuery(
                    """
                    select count(*)
                    from public.properties
                    where id = :propertyId and host_id = :hostId and deleted_at is null
                    """)
                .setParameter("propertyId", propertyId)
                .setParameter("hostId", hostId)
                .getSingleResult();
    return count.intValue() > 0;
  }

  @SuppressWarnings("unchecked")
  public List<HostBookingRow> listBookings(UUID hostId) {
    List<Object[]> rows =
        entityManager
            .createNativeQuery(
                """
                select
                  b.id,
                  b.status::text,
                  b.booking_type::text,
                  b.check_in,
                  b.check_out,
                  b.guest_count,
                  b.customer_notes,
                  p.title,
                  r.name,
                  s.total_krw,
                  b.created_at
                from public.bookings b
                join public.properties p on p.id = b.property_id
                join public.rooms r on r.id = b.room_id
                join public.booking_price_snapshots s on s.booking_id = b.id
                where p.host_id = :hostId
                order by b.created_at desc
                """)
            .setParameter("hostId", hostId)
            .getResultList();

    List<HostBookingRow> results = new ArrayList<>();
    for (Object[] row : rows) {
      results.add(
          new HostBookingRow(
              (UUID) row[0],
              BookingStatus.valueOf((String) row[1]),
              BookingType.valueOf((String) row[2]),
              toLocalDate(row[3]),
              toLocalDate(row[4]),
              ((Number) row[5]).intValue(),
              (String) row[6],
              (String) row[7],
              (String) row[8],
              ((Number) row[9]).intValue(),
              (OffsetDateTime) row[10]));
    }
    return results;
  }

  private static LocalDate toLocalDate(Object value) {
    if (value instanceof LocalDate localDate) {
      return localDate;
    }
    if (value instanceof java.sql.Date sqlDate) {
      return sqlDate.toLocalDate();
    }
    return LocalDate.parse(value.toString());
  }
}

File: apps/backend-node/src/features/hosts/host.repository.ts
import { Injectable } from '@nestjs/common';
import type { hosts } from '@prisma/client';

import { PrismaService } from '../../prisma/prisma.service';

export interface HostPropertyListRow {
  id: string;
  title: string;
  slug: string;
  propertyType: string;
  district: string;
  status: string;
  bookingMode: string;
  monthlyPriceMin: number | null;
  roomCount: number;
  updatedAt: Date;
}

export interface HostBookingRow {
  id: string;
  status: string;
  bookingType: string;
  checkIn: Date;
  checkOut: Date;
  guestCount: number;
  customerNotes: string | null;
  propertyTitle: string;
  roomName: string;
  totalKrw: number;
  createdAt: Date;
}

@Injectable()
export class HostRepository {
  constructor(private readonly prisma: PrismaService) {}

  async getByProfileId(profileId: string): Promise<hosts | null> {
    return this.prisma.hosts.findFirst({
      where: {
        profile_id: profileId,
        deleted_at: null,
      },
    });
  }

  async register(profileId: string, displayName: string): Promise<hosts> {
    const existing = await this.getByProfileId(profileId);
    if (existing) {
      return existing;
    }

    return this.prisma.$transaction(async (tx) => {
      await tx.$executeRaw`
        ALTER TABLE public.profiles DISABLE TRIGGER profiles_protect_role
      `;

      try {
        await tx.$executeRaw`
          UPDATE public.profiles
          SET role = 'host'
          WHERE id = ${profileId}::uuid
        `;

        return tx.hosts.create({
          data: {
            profile_id: profileId,
            display_name: displayName.trim(),
            status: 'pending',
          },
        });
      } finally {
        await tx.$executeRaw`
          ALTER TABLE public.profiles ENABLE TRIGGER profiles_protect_role
        `;
      }
    });
  }

  async listProperties(hostId: string): Promise<HostPropertyListRow[]> {
    const rows = await this.prisma.$queryRaw<
      {
        id: string;
        title: string;
        slug: string;
        property_type: string;
        district: string;
        status: string;
        booking_mode: string;
        monthly_price_min: number | null;
        room_count: number;
        updated_at: Date;
      }[]
    >`
      SELECT
        p.id,
        p.title,
        p.slug,
        p.property_type::text AS property_type,
        p.district,
        p.status::text AS status,
        p.booking_mode::text AS booking_mode,
        p.monthly_price_min,
        COALESCE(
          (
            SELECT COUNT(*)::integer
            FROM public.rooms r
            WHERE r.property_id = p.id
              AND r.deleted_at IS NULL
          ),
          0
        ) AS room_count,
        p.updated_at
      FROM public.properties p
      WHERE p.host_id = ${hostId}::uuid
        AND p.deleted_at IS NULL
      ORDER BY p.updated_at DESC
    `;

    return rows.map((row) => ({
      id: row.id,
      title: row.title,
      slug: row.slug,
      propertyType: row.property_type,
      district: row.district,
      status: row.status,
      bookingMode: row.booking_mode,
      monthlyPriceMin: row.monthly_price_min,
      roomCount: row.room_count,
      updatedAt: row.updated_at,
    }));
  }

  async propertyBelongsToHost(
    hostId: string,
    propertyId: string,
  ): Promise<boolean> {
    const count = await this.prisma.properties.count({
      where: {
        id: propertyId,
        host_id: hostId,
        deleted_at: null,
      },
    });

    return count > 0;
  }

  async listBookings(hostId: string): Promise<HostBookingRow[]> {
    const rows = await this.prisma.$queryRaw<
      {
        id: string;
        status: string;
        booking_type: string;
        check_in: Date;
        check_out: Date;
        guest_count: number;
        customer_notes: string | null;
        property_title: string;
        room_name: string;
        total_krw: number;
        created_at: Date;
      }[]
    >`
      SELECT
        b.id,
        b.status::text AS status,
        b.booking_type::text AS booking_type,
        b.check_in,
        b.check_out,
        b.guest_count,
        b.customer_notes,
        p.title AS property_title,
        r.name AS room_name,
        s.total_krw,
        b.created_at
      FROM public.bookings b
      JOIN public.properties p ON p.id = b.property_id
      JOIN public.rooms r ON r.id = b.room_id
      JOIN public.booking_price_snapshots s ON s.booking_id = b.id
      WHERE p.host_id = ${hostId}::uuid
      ORDER BY b.created_at DESC
    `;

    return rows.map((row) => ({
      id: row.id,
      status: row.status,
      bookingType: row.booking_type,
      checkIn: row.check_in,
      checkOut: row.check_out,
      guestCount: row.guest_count,
      customerNotes: row.customer_notes,
      propertyTitle: row.property_title,
      roomName: row.room_name,
      totalKrw: row.total_krw,
      createdAt: row.created_at,
    }));
  }

  async getProfileFullName(profileId: string): Promise<string | null> {
    const profile = await this.prisma.profiles.findFirst({
      where: {
        id: profileId,
        deleted_at: null,
      },
      select: { full_name: true },
    });

    return profile?.full_name ?? null;
  }
}
